Ical.Net/TimeZone/VTimeZoneExtensions.cs

```diff
@@ -0,0 +1,189 @@
+    public static CalDateTimeZone ToDateTimeZone(this VTimeZone vTimeZone, int untilYear)
+    {
+        if (TryCreateIntervals(vTimeZone, untilYear, out var tzId, out var intervals))
+            return new CalDateTimeZone(tzId, intervals);
+
+        throw new ArgumentException($"VTimeZone '{vTimeZone.TzId}' does not contain valid TimeZoneInfos.", nameof(vTimeZone));
+    }
+
+    /// <summary>
+    /// Converts a <see cref="T:System.DateTime" /> of any kind to a <see cref="LocalDateTime"/> in the ISO calendar.
+    /// Then, the given <paramref name="offset"/> is applied to the <see cref="LocalDateTime"/> to create an <see cref="Instant"/>.
+    /// <para/>
+    /// Any <see cref="P:System.DateTime.Kind" /> will create the same day/hour/minute etc.
+    /// </summary>
+    private static Instant CreateInstant(DateTime dt, Offset offset)
+    {
+        return LocalDateTime
+            .FromDateTime(dt)
+            .WithOffset(offset)
+            .ToInstant();
+    }
+
+    /// <summary>
+    /// A temporary class to hold transition information
+    /// that can be used to create <see cref="ZoneInterval"/>s.
+    /// </summary>
+    private sealed class Transition(string tzName, Instant start, Offset offset, Offset savings, Instant? end = null)
+    {
+        public string TzName { get; } = tzName;
+        public Instant Start { get; } = start;
+
+        /// <summary>
+        /// Will be set in <see cref="CreateIntervalsFromTransitions"/>
+        /// </summary>
+        public Instant? End { get; set; } = end;
+        public Offset Offset { get; } = offset;
+        public Offset Savings { get; } = savings;
+    }
+
+    private static bool TryCreateIntervals(VTimeZone vTimeZone, int untilYear, out string tzId, out List<ZoneInterval> intervals)
+    {
+        if(vTimeZone.TzId == null)
+            throw new ArgumentNullException(nameof(vTimeZone), "VTimeZone.TzId must not be null");
+
+        var transitions = new SortedList<Instant, Transition>();
+        intervals = new List<ZoneInterval>();
+        tzId = vTimeZone.TzId;
+
+        var timeZoneInfos = vTimeZone.TimeZoneInfos
+            .Where(tzi => tzi.Start != null)
+            .OrderBy(tzi => tzi.Start!.Value)
+            .ToList();
+
+        if (timeZoneInfos.Count == 0)
+            return false;
+
+        foreach (var tzi in timeZoneInfos)
+        {
+            var zoneTransitionsCount = 0;
+            var isDaylight = tzi.Name?.Equals(Components.Daylight, StringComparison.OrdinalIgnoreCase) == true;
+            var tzName = tzi.TimeZoneName ?? (isDaylight ? "DST" : "STD");
+            var offsetFrom = tzi.OffsetFrom != null ?  Offset.FromTimeSpan(tzi.OffsetFrom.Offset) : Offset.Zero;
+            var offsetTo = tzi.OffsetTo != null ? Offset.FromTimeSpan(tzi.OffsetTo.Offset) : Offset.Zero;
+            var savings = isDaylight ? (offsetTo - offsetFrom) : Offset.Zero;
+
+            foreach (var rrule in tzi.RecurrenceRules)
```

Comment:
The whole evaluation logic is implemented in `TimeZoneInfoEvaluator` (which doesn't add any functionality over `RecurringEvaluator` and could therefore be removed), so I recommend reusing that.

Reply:
Oh yes, sure - sorry